PR Cookbook

Andrew00000 opened the pull request:
Mi tetszik mi nem tetszik, te hogy csinalnad stb. Probaltam clean architecturet de ha szted nem az a legjobb v hasznosabb akkor nem kell h clean legyen. Inkabb legyen olyan ahogy te csinalnad :D

Commit:
Projects created

Commit:
Added come contracts

Added:
 - CreateRecipeRequest
 - UpdateRecipeRequest
 - RecipeResponse
 - RecipesResponses

Commit:
Added models

Added:
- Recipe
- Ingredient
- UnitType

Commit:
Added reference exclusively for DI

Commit:
Update UnitType.cs

Commit:
Update Recipe.cs

Commit:
Update Ingredient.cs

Commit:
Added in memory repository

Commit:
DI for reposittory

Commit:
Update Recipe.cs

Commit:
Create RecipesController.cs

Created `RecipesController` and added `CreateRecipeRequest`

Commit:
Update RecipesController.cs

Changed returned status code to 201

Commit:
Added contract mapping

Commit:
Clean up

Commit:
Removed magic strings

Added and API endpoints class to collect all the string for the APIs in one place.

Commit:
Added GetByID

Commit:
Added GetAllRecipes

Commit:
Clean up

Commit:
Added Update Recipe

Commit:
Added delete recipe

Commit:
Moved the models into a folder

Commit:
Implemented retrivel by slug

Commit:
Added GetAllTitles

Commit:
Added retrivel by given tag

Its possible to get back all of the recipes including a given tag

Commit:
Update UnitType.cs

Commit:
Update .gitignore

Commit:
Changed volume to amount

Reason: seemed more fitting

Commit:
Added packages to handles sqlite db

Commit:
Started db migration

Removed in memory db
Added sqlite db
Started implementing the new methods for communicating with the db.
Since the application isnt operational I removed the in memory db before I finished implementing  all of the methods for the new db.

Commit:
Added more methods working with the slug insted of the guid

more user friendly

Commit:
More DI

Commit:
Moved DbConnectionFactory to the Infrastructure project

I moved the implementation of IDbConnection to its own file and to the infrastructure project to comply with the clean architecure.

Commit:
Updated field name to NumberOfPortions

More descriptive

Commit:
Updated field name to NumberOfPortions

Commit:
Updated UnitType

Extended the list. plus by assigning values I can guarantee that in case of extending the unittype class, order change wont case an issue.

Commit:
Updated Recipe model

I made Author required because the slug is used for identification and it uses the Author.
Renamed Portions

Commit:
Added GetAllRecipes query

Commit:
GetAllAsync() uses db now

Commit:
Changed get All with tags to get all titles with tag

Result feels more useable

Commit:
Added Existance checks

Update will use the existance checks

Commit:
Corrected incorrect parameters

Commit:
Added Get functionalities with db + Small refactoring

Added GetByIdAsync and GetBySlugAsync
Added SqpliteCommandTexts for the Get methods
Extracted a couple methods for easier readablity

Commit:
Refactored GetAllTitles

Now it filters for the titles in the db, instead of creating the recipes and then only return the titles

Commit:
Update RecipesController.cs

Commit:
Refactored GetAllTitlesWithTag

Now it gets back a IEnumerable of titles from db instead of getting back an IEnumerable of recipes and return only the titles

Commit:
Tiny cleanup

Commit:
Added oredering to returned IEnumerables

It returns Recipes/Titles in alphabetical order, it is more user friendly

Commit:
Updated Endpoints to take id or slug instead of just id

Commit:
Implemented deletion by id and slug

Commit:
Implemented updating by id and slug

Commit:
Tiny clean up

Commit:
Update RecipesController.cs

Commit:
Introduced a business layer

The new business layer can have our application logic. Needed since neither the controller nor the repository should hold business logic.

Separated our services in a CQRS style, one interface for the commands/writes and another for the queries/reads.

For now most of our methods are just proxies in the business layer but when we have more logic we have the proper place to put it

Commit:
Update ContractMapping.cs

Commit:
Update RecipesResponse.cs

Commit:
Added validation for recipe

Commit:
Renamed cookbook to recipebook for consistency

Commit:
Added cancellation tokens

Commit:
Added cancellation tokens

Commit:
Update Cookbook.db

Andrew00000 committed:
Small resturcture

robert-fekete commented on Cookbook.API/ApiEndPoints.cs:
```diff
@@ -0,0 +1,20 @@
+﻿namespace Cookbook.API
+{
+    public static class ApiEndPoints
```
En amugy ilyen proper REST api-n sose dolgoztam cegnel, szoval ha valamivel ellent mondok a YouTube-os arcoknak, akkor lehet nekik van igazuk.

Ez pl egy pattern amit a videoban lattal, vagy te akartad szetvalasztani? Szerintem ez ront a dolgon inkabb mint segit. En nem akarnek kulon file-ba menni hogy lassam hogy mi az ertek.

Annak lehet ertelme hogy a Base-t kirakod egy const-ba a file-on belul, es akkor `$"{Base}/{{idOrSlug}}"`-t irsz a `[]`-be, de nem raknam ki oket kulon file-ba. Nem hasznalod egyiket tobbszor amugy vagy mas file-okbol, szoval en nem latok sok erteket egy kulon class-ban.

Andrew00000 replied:
Videoban volt. Talan az ilyen hosszabbaknal szebb "titles/tags/{tag}" de osszesegeben tell nem segit sokat. a Baset ki lehet rakni az elejere
```
namespace Cookbook.API.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class RecipesController : ControllerBase
```

robert-fekete commented on Cookbook.Application/Services/IRecipebookReadServices.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace Cookbook.Application.Services
+{
+    public interface IRecipebookReadServices
```
Miert plural? En singular-ban irnam

robert-fekete commented on Cookbook.Application/Services/RecipebookReadServices.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace Cookbook.Application.Services
+{
+    internal class RecipebookReadServices : IRecipebookReadServices
```
Szinten singular. Maga a class a service, nem 1-1 function

Andrew00000 replied:
ez zavart meg a nevezesben

robert-fekete commented on Cookbook.Application/Services/IRecipebookReadServices.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public interface IRecipebookReadServices
+    {
+        Task<IEnumerable<Recipe>> GetAllAsync(CancellationToken token = default);
```
Nekem nem tetszik hogy van default. Nem kene ezeket token nelkul hasznalni, es jobb ha a compiler enforce-olja.

robert-fekete commented on Cookbook.Application/Services/RecipebookWriteServices.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace Cookbook.Application.Services
+{
+    internal class RecipebookWriteServices : IRecipebookWriteServices
```
Miert valasztottad szet a read-write service-t?

Andrew00000 replied:
a Clean codes video ajanlotta aztan jo otletnek tunt igy van 2 api es ha mondjuk van 1 felhasznalo akinek nem kene tudnia irnia akkor ha csak a read apit kapja meg akkor veletlen se tud valtoztatni. Vagy nem kene? Kozben rajottem h az implementaciot nincs ertleme szetvalasztani :D vagy te is azt kerdezed?

robert-fekete replied:
igen igen igen. Az interface az jo otlet, eleg distinct responsibility-k meg lehet azt mondani hogy valaki esetleg csak olvasni akar, akkor kisebb "requirement" a read only interface, ez az I a SOLID-ban is, szoval az interface az okes, az implementacio lepett meg. Nem tudtam hogy "veletlen" vagy valahol olvastad hogy igy fasza

robert-fekete commented on Cookbook.Application/Validators/RecipeValidator.cs:
```diff
@@ -0,0 +1,44 @@
+            this.recipebookRepository = recipebookRepository;
+            
+            RuleFor(x => x.Title).NotEmpty();
+
```
NIT: az uj sorokat skip-pelnem

robert-fekete commented on Cookbook.Application/ApplicationServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace Cookbook.Application
+{
+    public static class ApplicationServiceCollectionExtensions
```
Ezt lattad valahol hogy igy kell? `ApplicationServiceCollectionExtensions`?

Andrew00000 replied:
Videoban csinalta igy a srac te hogy nevezned/csinalnad?

robert-fekete replied:
Pont igy, igazabol azert kerdeztem :D 2019-ben eleg uj volt meg ez az egesz dependency injection C#-ban es nem volt etul sok blog, video, javaslat rola. Elkezdtunk egy uj service-t a milestone-nal, aztan szetszedtuk tobb project-re es belefutottunk ebbe a problemaba hogy vagy minden public, implementatcio, interface, etc, vagy hogyan hook-olod fel az IServiceCollection-re. 
Akkor irtam egy one pager-t a problemarol, 3 opcioval es ez volt az egyik, meg emelett is dontottunk :D Csak meglepett hogy most ugyanezt viszont latom, de akkor ugy nez ki ez lett azota a best practice :D

Andrew00000 replied:
na jo bevallom igazabol olvastam a one pagered :D

Andrew00000 replied:
amugy azt hittem konkretan a nevre kerdezel ra h miert az :D nem arra h van kulon kis class :D

robert-fekete commented on Cookbook.Application/ApplicationServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,19 @@
+    {
+        public static IServiceCollection AddApplication(this IServiceCollection services)
+        {
+            services.AddValidatorsFromAssemblyContaining<IApplicationMarker>(
```
Ez eleg hacky nem? nem tudod inkabb a RecipeValidator-t hasznalni az IApplicationMarker helyett?

Andrew00000 replied:
na megneztem megint a videoban meg probaltam utana nezni. Csavesz ezt hasznalta de nem indokolta meg h miert. Mukodik RecipeValidatorral.

robert-fekete commented on Cookbook.Application/Persistance/IDbConnectionFactory.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Cookbook.Application.Database
+{
+    public interface IDbConnectionFactory
```
O miert ebben a projectben van? senki nem hasznalja itt, nem?

robert-fekete commented on Cookbook.Infrastructur/InfrastructureServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    public static class InfrastructureServiceCollectionExtensions
+    {
+        public static IServiceCollection AddPersistence(this IServiceCollection services)
```
Ezzel a Persistence kifejezessel meg sose talalkoztam kodban. Ismerem a kifejezest, meg ez tenyleg persisting, de nem lattam meg nevnek hasznalni. 

Repositories nem lenne jobb?

Andrew00000 replied:
nekem mindegy melyik :D COnal volt par helyen persistence onnan jott de repositories pontosabb

robert-fekete commented on Cookbook.Infrastructur/InfrastructureServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,23 @@
+
+        public static IServiceCollection AddDatabase(this IServiceCollection services)
+        {
+            services.AddSingleton<IDbConnectionFactory>(_ => new SqliteConnectionFactory());
```
miert nem `services.AddSingleton<IDbConnectionFactory, SqliteConnectionFactory>();`?

Andrew00000 replied:
csavo videoban igy irta de o at is adta a connection stringet amit en is fogok :D connection stringgel sem okes?

robert-fekete commented on Cookbook.Infrastructur/Persistance/DbConnectionFactory.cs:
```diff
@@ -0,0 +1,25 @@
+{
+    public class SqliteConnectionFactory : IDbConnectionFactory
+    {
+        private readonly string connectionString =
```
Ezt constructorban kene megkapnia. Nezz utana a Configuration-nek, van arra is tok jo support mind a ServiceCollection

robert-fekete commented on Cookbook.Infrastructur/Persistance/DbConnectionFactory.cs:
```diff
@@ -0,0 +1,25 @@
+
+        static SqliteConnectionFactory()
+        {
+            Batteries.Init();
```
batteries..., innit?

robert-fekete replied:
honnan jon amugy ez a Batteries vagy mi ez?

Andrew00000 replied:
lol :D

> batteries..., innit?

Andrew00000 replied:
> honnan jon amugy ez a Batteries vagy mi ez?

nem akart nelkule mukodni :D VS  "Show potential fixes" adta

robert-fekete replied:
marmint ugy ertem hogy ez milyen namespace vagy property vagy mi ez? milyen class?

robert-fekete replied:
Olyan mintha valami static class lenne, de nem latom melyik namespace alatt

robert-fekete replied:
SQLitePCL-bol jon :D nem tudom mi ez

Andrew00000 replied:
most kiszedtem es mukodik minden nelkule :D nem tom mit csinaltam h bekeveredett de akkor szenvedtem vele arra emlekszem lehet volt egy masik bugom ami kihozta es orvosolta vagy nem tom :D lehet lesz egy sunyi bugom most XD

robert-fekete commented on Cookbook.Infrastructur/Persistance/RecipebookRepository.cs:
```diff
@@ -0,0 +1,283 @@
+
+        public async Task<bool> CreateAsync(Recipe recipe, CancellationToken token)
+        {
+            using var connection = await dbConnectionFactory.CreateConnectionAsync(token);
```
Nem tudom hogy itt  melyik a helyes, minden keresre uj connection-t nyitni, vagy egyet hasznalni per repository. Ennek utana neztel esetleg?

Andrew00000 replied:
konkretan utana nem neztem, videoban meg COnal igy volt. de utananezek majd

Andrew00000 replied:
te egyet szoktal nyitni es hasznalni?

robert-fekete replied:
megmondom oszinten nem irtam meg ilyen kodot szerintem :/ altalaban ami service-eken dolgoztam ez a kod mar megvolt irva, pretty much az elso amit meg kell irnod, meg rohadt ritkan, kb soha nem valtozik. szoval nincs igazan tapasztalatom, nem tudnam megmondani hogy hogy szoktak/szoktuk. 
Ezert mondom hogy ha ezt lattad videoban vagy valahol, akkor biztos jo igy, de ha nem lattad sehol csak igy sikerult, akkor lehet meger egy kis research-ot, hogy hogy sokas. Ha videoban igy volt akkor jo ez igy szerintem.
Using az fontos, de az ott is van

Andrew00000 replied:
a using nem dobja el a vegen mindig a connectiont? es akkor kell uj kovi alkalomra

robert-fekete commented on Cookbook.Infrastructur/Persistance/RecipebookRepository.cs:
```diff
@@ -0,0 +1,283 @@
+                                new CommandDefinition(SqliteCommandTexts.GetAllTitles,
+                                                      cancellationToken: token));
+
+            if (titles is null)
```
Nem tudom melyik a szebb de tudsz ilyet is `return titles?.Order() ?? Enumerable.Empty<string>();`

Andrew00000 replied:
I like this better :D

robert-fekete commented on Cookbook.Infrastructur/Persistance/RecipebookRepository.cs:
```diff
@@ -0,0 +1,283 @@
+            var rawGuid = await connection.QuerySingleOrDefaultAsync(
+                       new CommandDefinition(SqliteCommandTexts.GetIdFromSlug,
+                                             new { slug }, cancellationToken: token));
+            return new Guid(rawGuid!.Guid);
```
Mi van ha nincs recipe ilyen sluggal a rendszerben?

Andrew00000 replied:
itt mi lenne a jo kezeles? throwing an exceptiont kicsit tulzasnak erzem de nem tom akkor mit kene visszaadni

robert-fekete replied:
`Guid?`

robert-fekete commented on Cookbook.Infrastructur/Persistance/SqliteCommandTexts.cs:
```diff
@@ -0,0 +1,121 @@
+﻿namespace Cookbook.Infrastructur.Persistance
+{
+    public static class SqliteCommandTexts //parameterize (create tables class, columns class)
```
a comment az meg mindig el? ez todo vagy mit jelent ez?

Andrew00000 replied:
ez TODO akart lenni, de akartalak kerdezni h van e ertelme. Arra gondoltam h csinalok 
egy 

public class Tables
{
public const string Recipe = "Recipe";
}
meg hozza
public class RecipeTable
{
public const string  Title = "Title";
public const string Author = "Author"
}

es akkor ha valamikor valtozik a string egy helyen van, meg ha irom a sql querriket akkor felajanlja a VS es nem irom el veletlen

v nincs ertelme/nem szoktak? a column nev valtoztatast tudom h ritka de COnal Axel ajanlotta h csinaljak ilyet

robert-fekete replied:
Nem, ez jo practice. Find reference is jobban mukodik mint a Ctrl + F. 
Azt nem talaltam meg pl hogy a tablakat hol hozod letre, akkor lenne igazan ertelme ha a create table-ben is tudod ujra hasznalni.

robert-fekete commented on Cookbook.Infrastructur/Persistance/RecipebookRepository.cs:
```diff
@@ -0,0 +1,283 @@
+using Dapper;
+using System.Data;
+
+namespace Cookbook.Infrastructur.Persistance //restructure DB to use rowID as FK instead of slug
```
//restructure DB to use rowID as FK instead of slug => yes please

Andrew00000 replied:
lol

robert-fekete commented on Cookbook.Infrastructur/InfrastructureServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,23 @@
+using Cookbook.Infrastructur.Persistance;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Cookbook.Infrastructur
```
Ez az Infrastructure nev is eleg fura nekem, foleg igy hogy ekkora a project. 
Infrastructure-aba bele tartoznak a serverek, az adatbazisok, a message queue-k, de az inkabb a "hosting" contextusban letezik, mi operations soran beszelunk infrastructure-rol, hogy hany hostot kell patch-elnunk, meg hogy kell uj host-okat rendelnunk.

Ha lenne message queue-d meg SNS meg nem tudom mi, akkor meg lehet mukodne a nev, de igy inkabb hivnam en Database-nek vagy Repositories-nak

Andrew00000 replied:
hey hey hey its Infrastructur

Andrew00000 replied:
but okay it makes sense

robert-fekete commented on Cookbook.Domain/Models/UnitType.cs:
```diff
@@ -0,0 +1,29 @@
+        cloves = 11,
+
+        tablespoon = 12,
+        tablespoons =13, 
```
Ezek a plural dolgok nekem nagyon furak. A unit az tablespoon, akar egy, akar ketto. A kifejezes, angolban, nyilvan tablespoons, de magyarban mar pl nem. Mi van ha magyar localization front end fut ugyanezzel a backend-del? Szerintem nem jo amikor egy adott nyelv iranyitja a model-t.

Andrew00000 replied:
de akkor h kezelned?

robert-fekete replied:
hat en singular-t hasznalnek az enum-ban, aztan egy kulon logika hogy ebbol hogy lesz UI string. De ezzel nem is kell igazan foglalkoznod mert nincs front end. A Json-ben is a unit szerintem csak egy "marker" kene legyen, aztan egy logika atkonvertalja majd arra amit a UI designer akar. 
- Magyarban nyilvan mas lesz a szo
- Magyarban nincs is tobbes szam, "2 teaskanalak"
- Lehet telefonon roviditeni akar, a hely miatt "tbsp"
- Lehet recept editorban akar egy dropdown-t ahol kivalaszthatod a unit-ot, ott nem jelenhet meg a singular meg a plural is vagy akkor kene valami elbaszott validator hogy az "1 tablespoons"-t ne fogadja el meg a "2 tablespoon"-t
- Lehet hogy akar valaki majd "scaling" funkcionalitast, hogy beirhasd hogy nem 2 hanem 4 vagy 8 fore akarod megfozni. Akkor az 1 tablespoon-bol ketto lesz, de az nem csak egy szorzassal jarna hanem a unit-ot is meg kene valtoztatnod singular-rol plural-ra
- Lehet hogy be lehet allitani preferred unit-ot, ahol automatikusan konvertalja a front end, akkor megint bonyolultabb ket enum erteket kezelni ugyanarra
- UI-wise maga az enum is inkonzisztens, mert most random lasagne receptekre ranezve, vagy konzisztensen roviditenek (tbsp, gr, ml) vagy konzisztensen kiirjak (tablespoon, grams, milliliter). Ez nem baj ha az enum teljesen decouple-d a UI-tol, azt irsz amit akarsz, de emiatt sem erdemes siman ToString-re bizni a UI string-et.

en pont ugy kezelnem mint a kg, g meg a tobbit. Technically ott is a plural a nyelvtanilag helyes amikor kiirod ("grams", "liters"). Szoval itt szerintem csak enum-ban azt kell jelolnod hogy milyen unit, aztan majd a front end kitalalja hogy azt hogyan kommunikalja le a user-nek.

robert-fekete commented on Cookbook.Domain/Models/UnitType.cs:
```diff
@@ -0,0 +1,29 @@
+
+        piece = 8,
+        pieces = 9,
+        clove = 10,
```
Ez a fokhagyma miatt? Nem lehet arra is piece-t hasznalni? 

Lehet jobb lenne ez valami string-kent mint enum-kent. Ez vegtelensegig bovulhet az ingredient-tol meg recept-tol fuggoen. Cup, slice, can, bottle, etc?

Andrew00000 replied:
de akkor hogy kezelned azt a reszt h ugyanaz legyen a string? alapbol onnan jott az enum otlete h nem akartam h legyen kg kilo kilogram kilogramm etc hanem egy kontrolalt verzio van csak

robert-fekete replied:
hat ez jo kerdes, ez kicsit inkabb UX design kerdes, mint szoftver tervezes kerdes. Itt alapvetoen a nagy UX kerdes, hogy kik editalhatjak a recepteket. 
1. Maga a UX "render" meg lokalizacio, amit fent emlitettem csak akkor tervezheto (pl roviditunk kis kepernyon, kiirjuk nagyon), ha elore tudjuk a lehetseges unit-okat. Ebben az esetben egy szuk admin kor tudna recepteket edit-elni, akikkel kapcsolatban all a szoftver csapat es tudjak igazitani a unit-okat (mint feature request) az alapjan hogy mire van szukseg. Ez gyakorlatilag az az irany amit most te kovetsz, mert az enum csak compile time valtoztathato. 
Ebben az iranyban sem kell amugy ragaszkodni az enum-hoz, de developer-ek adhatnak csak hozza uj unit-ot, akkor nem lesz kg/kilo/kilogram issue, de talan karbantarthatobb a kod. 
2. Masik opcio, hogy barki hozzaadhat ujat. Ebben az esetben lenne valami UI felulet ahol te definialhatod a unit-odat, meg esetleg megadhatsz konverziot mas mar letezo unit-hoz, hogy segitsd a rendszert atvaltas soran. Ez relative nagy szabadsag az oldalon, mert akkor tenyleg irhatok befottes uveget vagy barmit, cserebe nem fog mukodni a lokalizacio, mert nem tudja a fejleszto csapat release time hogy miket kene lefordittatni. Egyes szam/tobbes szam se igazan mukodik mert nem tudom mi a tobbes szama annak a random string-nek amit beirtal
3. Egy koztes megoldas, en valoszinu ebbe az iranyba mennek ha nekem kene UX-rol dontenem (ami mar regen rossz), ez valahol a ketto kozott lenne feluton. Lenne egy szerkeszto felulet es adatbazisban tartanank a hasznalhato unit-ok listajat, de csak adott permission-nel erheto el a felulet (i.e. csak admin-oknak). Igy a lokalizaciot elbukjuk tovabbra is, de ha egy "ismert" user bazis szerkeszti csak, akkor kicsit advance-abb field-eket is hozza lehet adatni veluk (singualr, plural, rovidites, etc), es akkor a UX render resze meg mukodhet. 

Long story short, az nem a te feladatod amikor adat tipusokat valasztasz hogy "lekorlatozd" a user-t idiota duplikaciok hozzaadasaban. Az maga egy feature, ha erre szukseg van, akkor front end-en le kell fejleszteni. Az adat modellednek az a dolga hogy mas fejlesztok idiotizmusatol vedjek a kodot, nem az hogy a userek idiotizmusatol vedjek az adatot. 

Szerintem az nem egy jo practice hogy ujra kell forditani a kodot ha hozza akarunk adni egy uj unit-ot, szoval en inkabb string-et hasznalnek es az adatbazisba raknek egy unit tablat. Az mar mas kerdes hogy ki tolti fel azt a tablat, lehet siman az init script ami letrehozza a tablat, feltolti ugyanezt a listat es kesz, read only, meg lehet user is aki feltolti, es akkor jon hozza valami feature ami ebben segit. 

Mi csinaltunk Saxo-nal amugy olyant, az szerintem jo practice volt, hogy ebben az esetben van egy unit lista az init script-ben, es ha jon egy feature request hogy kene "bottle" is, akkor siman hozzaadjuk az adatbazishoz live ES update-ljuk a script-et hogy legkozelebb amikor nullarol letrehozza az adatbazist akkor mar rakja bele egybol.

robert-fekete commented on Cookbook.Domain/Models/Ingredient.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public class Ingredient
+    {
+        public required int Amount { get; init; }
```
Ezt beszeltuk mar, de en nem igy modelleztem volna. Amount-nak a recipe-be kene mennie, nem akarnek 10 ingredient-et a Milk-hez, csak mert az egyik 1 deci a masik 2. Meg mi van ha van 1000 ml-es milk meg 1 literes milk?

robert-fekete commented on Cookbook.API/Mapping/ValidationMappingMiddleware.cs:
```diff
@@ -0,0 +1,38 @@
+
+        public async Task InvokeAsync(HttpContext context)
+        {
+            try
+            {
+                await next(context);
+            }
```
Nem teljesen vagom ezt a middleware rendszert, itt most tortenik validacio? olyan mintha csak tovabb adnad a kovetkezo middleware-nek validacio nelkul, nem?

Andrew00000 replied:
en ugy ertelmeztem a videobol h amikor tovabbadja akkor lefut a fluent validation es ha nem sikerul akkor validation exceptiont dob mm ez a nezt lesz a trigger

robert-fekete replied:
ezt beszeltuk messenger-en, de vegulis nem ez a class validal, ez csak elkapja meg kezeli a validation soran felmerulo exception-t. 
next-ben fut le "minden mas", a tobbi middleware, aztan a vegen a controller fuggvenye

robert-fekete commented on Cookbook.API/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        [HttpGet(ApiEndPoints.Recipes.Get)]
+        public async Task<IActionResult> Get([FromRoute] string idOrSlug,
```
En ezeket szetvalasztanam ket endpoint-ba, de lehet jobb igy

Andrew00000 replied:
a videoban egyben volt  de szet is szedhetem, viszont akkor h nezne ki v hogy valasztja szet?

```
[HttpGet(ApiEndPoints.Recipes.Get)]
        public async Task<IActionResult> Get([FromRoute] string slug,
                                                CancellationToken token)
        {
            var recipe =  await recipebookReadServices.GetBySlugAsync(slug, token);

            if (recipe is null)
            {
                return NotFound();
            }

            var response = recipe.MapToResponse();
            return Ok(response);
        }

[HttpGet(ApiEndPoints.Recipes.Get)]
        public async Task<IActionResult> Get([FromRoute] Guid id,
                                                CancellationToken token)
        {
            var recipe = await recipebookReadServices.GetByIdAsync(id, token);

            if (recipe is null)
            {
                return NotFound();
            }

            var response = recipe.MapToResponse();
            return Ok(response);
        }
```

es a kod majd eldonti h amit kapott az Guid v string? nem fogja megzavarni h a JSONban string a guid?

robert-fekete replied:
Hmmmm, igen en igy valasztanam szet, de jo kerdes, lehet megzavarja es ezert volt egyben. Probald ki esetleg, azzal is tanulsz, nem tudom hogy "felismeri-e" a guid-ot vagy lehet-e valahogy "priorizalni" hogy kezd a guid-dal, ha nem guid akkor menj a string-re.

Andrew00000 replied:
tudja :D jo helyre megy okos ez a c# :D

robert-fekete commented on Cookbook.API/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,117 @@
+                            ? request.MapToRecipe(id)
+                            : request.MapToRecipe(
+                                await recipebookReadServices
+                                            .GetIdFromSlugAsync(idOrSlug, token));
```
Mi van ha nincs recipe ezzel a sluggal?

Andrew00000 replied:
good catch. "Not found"-ed

Andrew00000 committed:
Refactoring - corrected naming

Andrew00000 committed:
Refactoring - removed default

better if the compiler enforces it

Andrew00000 committed:
Refactoring - merged Recipebook read and write service implementation

there is no need for two classes

Andrew00000 committed:
Update RecipeValidator.cs

Andrew00000 committed:
Refactoring - moved interface to proper place

Andrew00000 committed:
Refactoring - changed persistance

more fitting

Andrew00000 committed:
Update RecipebookRepository.cs

Andrew00000 committed:
Refactoring - renamed Infrasturcture to Repository

considering its small size it is more fitting

Andrew00000 committed:
Update Cookbook.API.sln

Andrew00000 committed:
Update Cookbook.API.csproj

robert-fekete commented on Cookbook.Application/Validators/RecipeValidator.cs:
```diff
@@ -0,0 +1,44 @@
+            RuleFor(x => x.NumberOfPortions).GreaterThan(0)
+                    .WithMessage("Number of portions have to be larger than 0");
+
+            RuleFor(x => x.Ingredients).NotEmpty();
```
Valamit adhatnal ehhez meg, mert "10 kg taters" formatumot var es a parsing logikadbol jon az exception, nem innen

Andrew00000 replied:
Ingredients validalas "1 piece of something" errorban h mi az expected format :Expected: 1 unit name"

modeled  Ingredient in JSON 
![image](https://github.com/Andrew00000/Cookbook/assets/62332857/9536bfc4-69ea-40dc-badf-7de4bcc539ed)

elkapja a request validator 
nem kell kulon parseolni
visszafele meg vagy marad a string vagy szetszedhetem modelbol

robert-fekete replied:
Nice nice

robert-fekete commented on Cookbook.Application/Validators/RecipeValidator.cs:
```diff
@@ -0,0 +1,44 @@
+        {
+            this.recipebookRepository = recipebookRepository;
+            
+            RuleFor(x => x.Title).NotEmpty();
```
Ami kicsit erdekes, es nem tudom hogy ezzel mit lehet kezdeni, de elobb hal el egy exception-on minthogy ide eljutna. Kiprobaltam hogy egy request-bol kivettem a title-t, es elso lepesben megprobalja deserializalni a CreateRecipeRequest-et, es az nem sikerul neki es dob egy hibat. Ha az sikerul neki utana hajtja ezt csak vegre (pl a greaterThan(0)-t). Szoval hiaba van itt NotEmpty minden property-re, nem innen jon a hiba, hanem a deserializalasbol.

robert-fekete replied:
Ja kozben rajottem hogy azert mert a CreateRecipeRequest-et (a bejovo JSON-t) egyaltalan nem validalod, csak a Recipe-t ami megy az adatbazisba. 

Valami input validation kellene szerintem. Nem tudom pontosan hogy hogy szednem szet, de a CreateRecipeRequest-ben kene validalni hogy a NumberOfPortions az tenyleg int, a Title property tenyleg letezik, az Ingredients tenyleg egy array, stb. 
Most mindenfele elbaszott json-t bekuldhetek es vagy internal exception-t kapok (int.Parse elhal) vagy mar a deserialization is elhal. Egyik sem nice response. 
Aztan a mostani validation maradhat, ez inkabb ilyen business logic validation, hogy a number of portion nem zero, meg nem ures a title, a masik meg lehetne ilyen basic input validation hogy a tipusok stimmelnek meg a property-k megvannak.

Andrew00000 replied:
![image](https://github.com/Andrew00000/Cookbook/assets/62332857/bf24f7f6-956a-43b0-b251-a97030f8b52c)
en ezt kapom vissza ha kihagyom a titlet vagy ha mondjuk intet rakok bele. vagy ez nem jo? itt nem biztos h ertem mi a baj

Andrew00000 replied:
beszeltuk de ezt egyelore elengedtem kene hozza custombinding majd ha kesz visszanezek ra

robert-fekete commented on Cookbook.API/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        [HttpPost(ApiEndPoints.Recipes.Create)]
+        public async Task<IActionResult> Create([FromBody] CreateRecipeRequest request,
```
Ezt beszeltuk messenger-en de meg kene probalni elkapni azokat a validation error-okat amik akkor keletkeznek amikor a body-bol megprobalja kiolvasni a request-et

Andrew00000 committed:
Create Cookbook.postman_collection.json

Andrew00000 committed:
Removed ApiEndPoints

After some research it seems unnecessary.

Andrew00000 committed:
Removed IApplicationMarker

Did some research seems unnecessary

Andrew00000 committed:
Update .gitignore

Andrew00000 committed:
Refactoring - Added db folder

Andrew00000 committed:
Update InfrastructureServiceCollectionExtensions.cs

Andrew00000 committed:
Update .gitignore

Andrew00000 committed:
Update DbConnectionFactory.cs

Andrew00000 committed:
Changed FK from slug to rowId

Makes much much more sense :D

Andrew00000 committed:
Update RecipebookRepository.cs

Andrew00000 committed:
Namespace cleanup

Andrew00000 committed:
Refactored sql commands

Stored the table and column names as constants. Autofill works now plus reduced the chance of typo caused bugs

Andrew00000 committed:
Stored connection string in config json

Andrew00000 committed:
Refactoring - Splitted Get by id or slug to Get by Id and Get by slug

Andrew00000 committed:
Refactoring - Collected the tables in a designated folder

Andrew00000 committed:
Update RecipebookRepository.cs

Andrew00000 commented:
Create db init process

Andrew00000 committed:
Update RecipesController.cs

Andrew00000 committed:
Update ValidationMappingMiddleware.cs

Andrew00000 committed:
Refactoring - corrected naming

Andrew00000 committed:
Refactoring - replaced Guid with Recipe row id

Andrew00000 committed:
Update Cookbook.db

Andrew00000 committed:
Refactoring - changed how I update the db

This way the id wont change with every update.

Andrew00000 committed:
Split the update controller into 2

Andrew00000 committed:
Split the delete controller into 2

Andrew00000 committed:
Update RecipesController.cs

Andrew00000 committed:
Update Cookbook.db

Andrew00000 committed:
Refactoring

Removed UnitType, it would be easier to control that the units are uniform from the frontend and it shouldnt effect the design of backend

Andrew00000 committed:
Refactoring - Added DbManipulator

Added db manipulator as a defensive measure against someone forgetting "using" and causing issues.

Andrew00000 committed:
Refactoring - Added DatabaseConfiguration

Andrew00000 committed:
Update SqliteCommandTexts.cs

Andrew00000 committed:
Update CreateRecipeRequest.cs

Andrew00000 committed:
Refactoring - JSON uses ingredients model now

Andrew00000 committed:
Update Cookbook.postman_collection.json

Andrew00000 committed:
Update Cookbook.db

Andrew00000 committed:
Refactoring - changed the format of ValidationResponse

Now it makes sense even if we dont have a property name

Andrew00000 committed:
Added tests for the requests

Andrew00000 committed:
[Refactoring] Ingredients are returned alphabetically

Andrew00000 committed:
[Refactoring[Bugfix] Ingredients with space in their name now returned correctly

Andrew00000 committed:
[Refactoring] Unified error handling WIP

Andrew00000 committed:
[Refactoring][WIP] Error handling

Andrew00000 committed:
[Refactoring] Unified error responses

Andrew00000 committed:
[Refactoring] Added a db initializer

if the db file does not exists it will create it with the appropriate tables.

Andrew00000 commented:
> Create db init process

@robert-fekete pls check h milyen lett :D

Andrew00000 committed:
Update DbInitializer.cs

Andrew00000 committed:
[Refactoring] Code Clean Up

Andrew00000 committed:
Added JWT generator API

Obviously in a real service it would not be like this :) but for convenience sake I added one

Andrew00000 committed:
Update launchSettings.json

Andrew00000 committed:
Token generation works

Andrew00000 committed:
[Refactoring] Async correction

Andrew00000 committed:
[Refactoring] Folder rename

Andrew00000 committed:
Update DbInitializer.cs

Andrew00000 committed:
[Refactor] Now the connection is opened in a using block

Andrew00000 committed:
[Refactor] Introduced IInitializable

Andrew00000 committed:
[Refactor] Splitting recipe retrieval into separate queries per table
